File: js/Chess/BoardParser.js
import pieceColor from './PieceColor'
import Piece from './Piece'
import PieceType from './PieceType'
import Board from './Board'

export default class BoardParser {
    parseBoard(lines) {
        if (lines.length !== 8) {
            throw new Error('Should be exactly 8 lines');
        }

        if (lines.some(line => line.length !== 8)) {
            throw new Error('All lines should have 8 chars length');
        }

        const cells = [];
        for (let y = 0; y < 8; y++) {
            const line = lines[y];
            if (!line) {
                throw new Error('incorrect input');
            }
            cells[y] = [];
            for (let x = 0; x < 8; x++) {
                cells[y][x] = this.parsePiece(line[x]);
            }
        }
        return new Board(cells);
    }

    parsePiece(pieceSign) {
        const color = /[A-Z]/.test(pieceSign) ? pieceColor.white : pieceColor.black;
        const pieceType = this.parsePieceType(pieceSign.toUpperCase());
        return pieceType === null ? null : new Piece(pieceType, color);
    }

    parsePieceType(sign) {
        switch (sign) {
            case 'R': return PieceType.Rook;
            case 'K': return PieceType.King;
            case 'N': return PieceType.Knight;
            case 'B': return PieceType.Bishop;
            case 'Q': return PieceType.Queen;
            case ' ':
            case '.': return null;
            default: throw new Error("Unknown chess piece " + sign);
        }
    }
}

File: cs/Chess/BoardParser.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Chess
{
    public class BoardParser
    {
        private const int MaxLinesLength = 8;
        
        private static Dictionary<char, PieceType> charToPieceTypesConverter;
        public BoardParser()
        {
            charToPieceTypesConverter = new Dictionary<char, PieceType>()
            {
                { 'R', PieceType.Rook },
                { 'K', PieceType.King },
                { 'N', PieceType.Knight },
                { 'B', PieceType.Bishop },
                { 'Q', PieceType.Queen },
                { ' ', null },
                { '.', null },
            };
        }
        public Board ParseBoard(string[] lines)
        {
            if (lines.Length != MaxLinesLength) throw new ArgumentException("Should be exactly 8 lines");
            if (lines.Any(line => line.Length != MaxLinesLength)) throw new ArgumentException("All lines should have 8 chars length");

            var cells = new Piece[MaxLinesLength][];
            for (var y = 0; y < MaxLinesLength; y++)
            {
                var line = lines[y];
                if (line == null) throw new Exception("incorrect input");
                cells[y] = new Piece[MaxLinesLength];
                for (var x = 0; x < MaxLinesLength; x++)
                    cells[y][x] = ParsePiece(line[x]);
            }
            return new Board(cells);
        }

        private static Piece ParsePiece(char pieceSign)
        {
            var color = char.IsUpper(pieceSign) ? PieceColor.White : PieceColor.Black;
            var pieceType = ParsePieceType(char.ToUpper(pieceSign));
            return pieceType == null ? null : new Piece(pieceType, color);
        }

        private static PieceType ParsePieceType(char sign)
        {
            if(!charToPieceTypesConverter.ContainsKey(sign))
                throw new FormatException("Unknown chess piece " + sign);
            return charToPieceTypesConverter[sign];
        }
    }
}